### PR TITLE
Support vectorized `like` and `regexp`

### PR DESCRIPTION
**Pull request by wangshuo128**
## Proposed changes

- Support `VExprContext` and vectorized `FunctionContext`
- Implement  vectorized `like` and `regexp` based on vectorized `FunctionContext`

### Motivation
The therm `VExprContext` is just like `ExprContext` in scalar running mode.
The purpose of `VExprContext` is to evaluate an expression (tree)  concurrently. One specific expression (tree) could have at least one `VExprContext`. We should **clone** `VExprContext` whenever we want to evaluate an expression (tree) in a fresh new thread.

### Key points
The `VExprContext` holds the runtime state of its expression in one thread, such as `RuntimeState` and `FunctionContext.` 
`FunctionContext` holds the function-call related states, i.e., function arguments info (argument number/type/constant value, etc), function states, etc.

There are two kinds of function states:
- **fragment-local state**: immutable states shared in all `FunctionContext`s, i.e., pre-processed constant arguments.
- **thread-local state:** mutable states (not thread-safe) states that are isolated among multiple threads.

### fragment-local  state initialization
When a `VExprContext` is opened for the first time (in `prepare`  phase of `ExecNode`), `VectorizedFnCall::open` is called. Function`s fragment-local states and thread-local states are both initialized.

### thread-local state initialization
When we want to evaluate an expression (tree) in a new thread, we should clone a new `VExpreContext` for the expression (tree). Fragment-local states are copied by the `VExpr/Function` framework, thread-local states should be initialized by the specific function implementation itself.

The main code path is like below:

- `VExprContext::clone` 
Create a new `VExprContext`, call `FunctionContextImpl::clone` to create new `FunctionContext`, only copying argument info and **frgament-local** function states.
- `VectorizedFnCall::open` (thread-local state)
Delegate to `IFunction::prepare` with **thread-local** function states.
- `IFunction::prepare`
Initialize **thread-local** function states.

## Todo
BE would crash when we run `like`/`regexp` with a constant null pattern, i.e, `SELECT * from t WHERE value like null`. It seems that the current vectorized function-call framework cannot handle constant null function arguments correctly. I'd dig into this and try to send a following PR to fix this issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/exec/exec_node.cpp [hunk 1]
```diff
@@ -181,8 +181,9 @@ Status ExecNode::init(const TPlanNode& tnode, RuntimeState* state) {
         _vconjunct_ctx_ptr.reset(new doris::vectorized::VExprContext*);
         RETURN_IF_ERROR(doris::vectorized::VExpr::create_expr_tree(_pool, tnode.vconjunct,
                                                                    _vconjunct_ctx_ptr.get()));
+    } else {
```
Comment (HappenLee) [inline be/src/exec/exec_node.cpp:184]: 这里加else会导致谓词下推存储层的结果不对的，后面一个语句一定要执行的

File: be/src/vec/functions/function_helpers.cpp [hunk 2]
```diff
@@ -93,4 +93,30 @@ const ColumnConst* check_and_get_column_const_string_or_fixedstring(const IColum
     return {};
 }
 
+
+
+NullPresence getNullPresense(const Block& block, const ColumnNumbers& args) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_helpers.cpp:98]: 这个函数不是本身在function.cpp中就有吗？为什么要写两次呢？还有辛苦一起改为下划线命名法

File: be/src/vec/functions/like.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+    Status execute_impl(Block& block, const ColumnNumbers& arguments, size_t result,
+                        size_t /*input_rows_count*/) override {
+        // values and patterns
+        const auto values_col =
+                block.get_by_position(arguments[0]).column->convert_to_full_column_if_const();
+        const auto pattern_col =
+                block.get_by_position(arguments[1]).column->convert_to_full_column_if_const();
+
+        const auto* values = check_and_get_column<ColumnString>(values_col.get());
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:54]: use_default_implementation_for_nulls 设置为false了。
这里有可能是nullable列的，我的理解这个是不是应该默认走向量化本身的null处理逻辑就行了，use_default_implementation_for_nulls 应该为true

File: be/src/vec/functions/function_helpers.cpp (hunk 2 above)
Comment (wangshuo128) [reply be/src/vec/functions/function_helpers.cpp:98]: 原来只能在functions.cpp内部用，现在放到了function_helpers.h里面，作为工具使用。
我改一下命名。

File: be/src/vec/exprs/vexpr.h [hunk 4]
```diff
@@ -98,6 +127,16 @@ class VExpr {
...
+
+    /// Index to pass to ExprContext::fn_context() to retrieve this expr's FunctionContext.
+    /// Set in RegisterFunctionContext(). -1 if this expr does not need a FunctionContext and
+    /// doesn't call RegisterFunctionContext().
+    int _fn_context_index;
+
+    // If this expr is constant, this will store and cache the value generated by
+    // get_const_col()
+    std::shared_ptr<ColumnPtrWrapper> _constant_col;
+    std::shared_ptr<Block> _constant_block;
```
Comment (HappenLee) [inline be/src/vec/exprs/vexpr.h:139]: 这个指针感觉没什么必要，直接栈上用一个block，用完析构了是不是就完事了

File: be/src/vec/exprs/vexpr.h [hunk 5]
```diff
@@ -40,16 +42,32 @@ class VExpr {
...
-    virtual Status execute(vectorized::Block* block, int* result_column_id) = 0;
+
+    Status open(RuntimeState* state, VExprContext* context) {
+        return open(state, context, FunctionContext::FRAGMENT_LOCAL);
+    }
+
+    virtual Status open(RuntimeState* state, VExprContext* context,
+                        FunctionContext::FunctionStateScope scope);
+
+    virtual Status execute(VExprContext* context, vectorized::Block* block,
```
Comment (HappenLee) [inline be/src/vec/exprs/vexpr.h:53]: 这个参数为什么不放最后一个，然后补充一个默认参数VExprContext* context = nullptr,这样很多地方就不需要手动给补nullptr参数了

Comment (wangshuo128) [reply be/src/vec/exprs/vexpr.h:53]: 这个当时并没有想这么多。
但是担心做成default parameter，用的时候掉坑里。因为整个Expr, Function(尤其是函数)的继承体系比较复杂。
https://stackoverflow.com/questions/3533589/can-virtual-functions-have-default-parameters。

File: be/src/vec/columns/column.h [hunk 6]
```diff
@@ -437,3 +437,13 @@ bool is_column_const(const IColumn& column);
 bool is_column_nullable(const IColumn& column);
 
 } // namespace doris::vectorized
+
+namespace doris {
+struct ColumnPtrWrapper {
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/vec/columns/column.h:442]: 因为`ColumnPtr`没法forward declaration，我稍后加个注释。

File: be/src/udf/udf_ir.cpp [hunk 7]
```diff
@@ -25,13 +25,28 @@ bool FunctionContext::is_arg_constant(int i) const {
...
     }
     return _impl->_constant_args[i];
 }
 
+doris::ColumnPtrWrapper* FunctionContext::get_constant_col(int i) const {
+    if (i < 0 || i >= _impl->_constant_cols.size()) {
+        return nullptr;
+    }
+    return _impl->_constant_cols[i];
+    return nullptr;
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/udf/udf_ir.cpp:47]: 确实是，我再提个PR把这个删了。

File: be/src/vec/exprs/vexpr.h (hunk 5 above)
Comment (HappenLee) [reply be/src/vec/exprs/vexpr.h:53]: 这个问题应该不大把，如果没有默认参数，编译期就会报错了，问题也就明确了

File: be/src/vec/exprs/vexpr.h (hunk 4 above)
Comment (wangshuo128) [reply be/src/vec/exprs/vexpr.h:139]: 并不是所有Expr都是constant，只有是常量的Expr才需要set这部分

Comment (HappenLee) [reply be/src/vec/exprs/vexpr.h:139]: 我的意思是，是不是没必要保留这个成员变量，它只是计算的时候临时用一下，后续就没用了

Comment (wangshuo128) [reply be/src/vec/exprs/vexpr.h:139]: 这是个cache，避免重复运算，可以看 `VExpr::get_const_col`

File: be/src/vec/exprs/vexpr.h (hunk 5 above)
Comment (HappenLee) [reply be/src/vec/exprs/vexpr.h:53]: 这个还是建议放到后面加默认参数，每个函数有个nullptr有点太丑了

File: be/src/udf/udf_ir.cpp [hunk 8]
```diff
@@ -25,13 +25,27 @@ bool FunctionContext::is_arg_constant(int i) const {
     return _impl->_constant_args[i] != NULL;
 }
 
+bool FunctionContext::is_col_constant(int i) const {
+    if (i < 0 || i >= _impl->_constant_cols.size()) {
+        return false;
+    }
+    return _impl->_constant_cols[1] != nullptr;
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/udf/udf_ir.cpp:32]: Oops... 我改一下

File: be/src/vec/exprs/vectorized_fn_call.cpp [hunk 9]
```diff
@@ -47,30 +49,66 @@ doris::Status VectorizedFnCall::prepare(doris::RuntimeState* state,
...
+
+    FunctionContext::TypeDesc return_type = AnyValUtil::column_type_to_type_desc(_type);
+    std::vector<FunctionContext::TypeDesc> arg_types;
+    bool char_arg = false;
+    for (int i = 0; i < _children.size(); ++i) {
+        arg_types.push_back(AnyValUtil::column_type_to_type_desc(_children[i]->type()));
+        char_arg = char_arg || (_children[i]->type().type == TYPE_CHAR);
+    }
+
+    _fn_context_index = context->register_func(state, return_type, arg_types, 0);
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/vec/exprs/vectorized_fn_call.cpp:61]: `FunctionContext`属于`VExprContext`,  `VExprContext::clone`的时候，需要clone `VExprContext`里面的fragment local的state。

File: be/src/vec/exprs/vectorized_fn_call.cpp [hunk 10]
```diff
@@ -47,30 +49,66 @@ doris::Status VectorizedFnCall::prepare(doris::RuntimeState* state,
                 fmt::format("Function {} is not implemented", _fn.name.function_name));
     }
     _expr_name = fmt::format("{}({})", _fn.name.function_name, child_expr_name);
+
+    FunctionContext::TypeDesc return_type = AnyValUtil::column_type_to_type_desc(_type);
+    std::vector<FunctionContext::TypeDesc> arg_types;
+    bool char_arg = false;
```
[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/vec/exprs/vectorized_fn_call.cpp]: done

File: be/src/vec/exprs/vectorized_fn_call.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/vec/exprs/vectorized_fn_call.cpp:59]: VExprContext也可以持有这个指针，二者不冲突啊

[the remark this answers is not in the document]
Comment (wangshuo128) [reply be/src/vec/exprs/vectorized_fn_call.cpp:59]: 需要理解这一点：一个expr

File: be/src/vec/functions/like.cpp [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+            i += 2;
+        } else {
+            search_string->append(1, tmp_search_string[i]);
+            i++;
+        }
+    }
+}
+
+Status FunctionLike::prepare(FunctionContext* context, FunctionContext::FunctionStateScope scope) {
+    if (scope != FunctionContext::THREAD_LOCAL) {
```
Comment (HappenLee) [inline be/src/vec/functions/like.cpp:212]: 这个为什么只有thread_local的时候进行prepare呢？

File: be/src/vec/functions/like.h [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+    LikeSearchState search_state;
+    LikeFn function;
+};
+
+class FunctionLikeBase : public IFunction {
+public:
+    size_t get_number_of_arguments() const override { return 2; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& /*arguments*/) const override {
+        return make_nullable(std::make_shared<DataTypeUInt8>());
```
Comment (HappenLee) [inline be/src/vec/functions/like.h:81]: 这个函数不应该是always nullable的吧？

**Pull request merged**